### DISCUSSION TITLE
How to add space between a list and vertical scrollbar?

### DISCUSSION THREAD

**Discussion by byorgey** [jtdaugherty/brick#484] [General]
I am rendering a brick list and wanted to add a scrollbar.  That worked fine, but the scrollbar was immediately next to the list items with no intervening space, which I thought looked bad.  But since the `viewport` is actually created inside the `renderListWithIndex` function, it seems like I can't insert padding in between the rendered items and the `viewport` (without duplicating `renderListWithIndex` and making my own modified version).  I was able to create some space by putting padding on the end of the rendered list items, but now the highlighting for the current list item extends all the way to the scrollbar (see the screenshot below), which I don't want.  I just want a one-space-wide vertical blank strip in between the list items and the scrollbar.

Any ideas on how this might best be accomplished?

![list-viewport](https://github.com/jtdaugherty/brick/assets/533859/fc99909e-a5a2-4790-be2a-54c89d35291b)

**Comment by jtdaugherty** [comment]
Right now, I think the approach you took is the only way: to add a bit of padding to the right side of each list element. The undesirable attribute behavior can probably be fixed with some clever attribute map changes, but that isn't how I'd like this to work. I'd like to think about this some more, but I'm leaning toward augmenting the library's scroll bar configuration API with a spacing setting to make this a first-class feature.

**Reply by jtdaugherty** [reply]
Another idea I had was to make it so that you could just augment the scroll bar renderer itself, since that is already exposed. I think that's a little more indirect, but at least that hitches onto an existing abstraction rather than adding an API feature *only* for this spacing. Right now the viewport scrollbar rendering internals make the assumption that the scroll bar renderer (which you can already override) always and only produces something exactly one row or column in size. If I relax that assumption, then it becomes possible to make a scroll bar renderer that just adds padding where you want it.

**Reply by byorgey** [reply]
Thanks for the quick reply!  Yes, some kind of spacing setting(s) associated to the scroll bar seems like it could make sense.  In the meantime I can just live with a little bit of extra blue highlighting.

**Reply by byorgey** [reply]
Yes, I also looked into overriding the scroll bar rendering but realized that wouldn't let me do what I wanted. Allowing more control over the scrollbar rendering would definitely be a more general way to solve the problem.

**Reply by jtdaugherty** [reply]
Okay. I'll let you know when I have some progress to report!

**Reply by jtdaugherty** [reply]
Looking into this a bit, there is a good reason why I did it as I did: the scroll bar size cannot be dynamic, since it depends on the dimensions of the viewport contents. And the available space in the viewport for the contents depends on the size of the scrollbar. That only works if the scrollbar's declared presence (by the user) determines the scroll bar's dimensions (i.e. always exactly one row/column in size). So I think the only way to do this would be to add a scrollbar spacing notion to the scrollbar configuration API. I could add something to the list itself, but that would be specific to lists and I have even already run into cases myself where I wanted scrollbar spacing in non-list-based viewports.

**Reply by jtdaugherty** [reply]
@byorgey I've pushed 324b5adfbd8a0b7f345c5aeda0d95c5c11e1b280 to `master`. Please check it out! You can see a working example of a modified scroll bar renderer in `programs/ViewportScrollbarsDemo.hs`. In your case it would be something like the following, provided you wanted to piggyback on the default renderer without repeating it:
```haskell
addLeftSpacing :: VScrollbarRenderer n -> VScrollbarRenderer n
addLeftSpacing r =
    r { scrollbarWidthAllocation = 2
      , renderVScrollbar = hLimit 1 $ renderVScrollbar r
      , renderVScrollbarTrough = hLimit 1 $ renderVScrollbarTrough r
      }
```
and then
```haskell
myWidget :: Widget n
myWidget =
    ...
    withVScrollBarRenderer (addLeftSpacing verticalScrollbarRenderer) $
    ....
```
That will automatically result in spacing on the left if your scroll bar orientation is on the right.

**Reply by jtdaugherty** [reply]
After discovering a wart in the API, I went a bit further and introduced some more changes in https://github.com/jtdaugherty/brick/commit/6b7195cf6e5bc95f9f3f588b0ec68103c0c5cd6d that should be easy to adjust for, but I wanted to call that out in case you pull `master` later and get both changes. I'll edit my example above to account for the change.

**Reply by byorgey** [reply]
I tried it out and it works great!

**Reply by jtdaugherty** [reply]
Okay, great. I'll make a release over the weekend.

**Reply by jtdaugherty** [reply]
The changes are now released in 1.10.
